tests/torchtune/models/phi3/test_lora_phi3.py

```diff
@@ -0,0 +1,302 @@
+            lora_alpha=ALPHA,
+        )
+        fixed_init_model(lora_phi_sa)
+        return lora_phi_sa
+
+    def test_empty_lora_modules(self):
+        with pytest.raises(ValueError, match="Must pass one or more of"):
+            _ = self.get_lora_phi_self_attention([])
+
+    @pytest.mark.parametrize(
+        "lora_modules, expected",
+        [
+            (["q_proj", "v_proj"], torch.tensor(50.64567)),
+            (["q_proj", "k_proj", "v_proj", "output_proj"], torch.tensor(77.17097)),
+            (["k_proj"], torch.tensor(44.90090)),
+        ],
+    )
+    def test_forward(self, inputs, lora_modules, expected):
+        lora_phi_sa = self.get_lora_phi_self_attention(lora_modules)
+        actual = lora_phi_sa(inputs)
+        assert_expected(actual.shape, (BSZ, SEQ_LEN, EMBED_DIM))
+        assert_expected(actual.mean(), expected, atol=1e-4, rtol=1e-6)
+
+
+class TestLoRAPhi3:
+    @pytest.fixture
+    def vocab_size(self):
+        return 50
+
+    @pytest.fixture
+    def inputs(self, vocab_size):
+        return torch.randint(low=0, high=vocab_size, size=(BSZ, SEQ_LEN))
+
+    def get_lora_phi3(
+        self,
+        lora_modules,
+        apply_lora_to_mlp,
+        apply_lora_to_output,
+        vocab_size,
+        reset_norm=True,
+        quantize_base=False,
+        embed_dim=EMBED_DIM,
+        dtype=None,
+    ):
+        num_layers = 3
+        model = lora_phi3(
+            lora_attn_modules=lora_modules,
+            apply_lora_to_mlp=apply_lora_to_mlp,
+            apply_lora_to_output=apply_lora_to_output,
+            vocab_size=vocab_size,
+            num_layers=num_layers,
+            num_heads=NUM_HEADS,
+            num_kv_heads=NUM_KV_HEADS,
+            embed_dim=embed_dim,
+            intermediate_dim=INTER_DIM,
+            max_seq_len=MAX_SEQ_LEN,
+            lora_rank=RANK,
+            lora_alpha=ALPHA,
+            quantize_base=quantize_base,
+        )
+        # To make final outputs less trivial
+        if reset_norm:
+            model.norm = nn.Identity()
+
+        # dtype=None means to just read dtype from parameters
+        # in the model. This dtype is set explicitly to bf16 currently
+        # when initializing QLoRA models, as ops such as `arange` aren't
+        # yet supported with the actual nf4 tensor dtype yet.
+        fixed_init_model(model, dtype=dtype)
+
+        return model
+
+    def get_ref_phi3(self, vocab_size, embed_dim=EMBED_DIM):
+        num_layers = 3
+        model = phi3(
+            vocab_size=vocab_size,
+            num_layers=num_layers,
+            num_heads=NUM_HEADS,
+            num_kv_heads=NUM_KV_HEADS,
+            embed_dim=embed_dim,
+            intermediate_dim=INTER_DIM,
+            max_seq_len=MAX_SEQ_LEN,
+        )
+        return model
+
+    @pytest.mark.parametrize(
+        "lora_modules, apply_lora_to_mlp, apply_lora_to_output, expected",
+        [
+            (["q_proj", "v_proj"], False, False, torch.tensor(2869701.75)),
```

Comment:
same q about these tensor values - where do we get them from and how can a user reproduce?